compiler/noirc_evaluator/src/ssa/interpreter/errors.rs

```diff
@@ -0,0 +1,105 @@
+    BlockMissingTerminator { block: BasicBlockId },
+    #[error("constrain {lhs_id} == {rhs_id} failed:\n    {lhs} != {rhs}")]
+    ConstrainEqFailed { lhs: String, lhs_id: ValueId, rhs: String, rhs_id: ValueId },
+    #[error("constrain {lhs_id} != {rhs_id} failed:\n    {lhs} == {rhs}")]
+    ConstrainNeFailed { lhs: String, lhs_id: ValueId, rhs: String, rhs_id: ValueId },
+    #[error(
+        "Range check of {value_id} = {value} failed.\n  Max bits allowed by range check = {max_bits}\n  Actual bit count = {actual_bits}"
+    )]
+    RangeCheckFailed { value: String, value_id: ValueId, actual_bits: u32, max_bits: u32 },
+    #[error(
+        "Range check of {value_id} = {value} failed.\n  Max bits allowed by range check = {max_bits}\n  Actual bit count = {actual_bits}\n  {message}"
+    )]
+    RangeCheckFailedWithMessage {
+        value: String,
+        value_id: ValueId,
+        actual_bits: u32,
+        max_bits: u32,
+        message: String,
+    },
+    #[error("Call to unknown foreign function {name}")]
+    UnknownForeignFunctionCall { name: String },
+    #[error("Cannot call non-function value {value_id} = {value}")]
+    CalledNonFunction { value: String, value_id: ValueId },
+    // Note that we don't need to display the value_id because displaying a reference
+    // value shows the original value id anyway
+    #[error("Reference value `{value}` passed from a constrained to an unconstrained function")]
+    ReferenceValueCrossedUnconstrainedBoundary { value: String },
+    #[error("Reference value `{value}` loaded before it was first stored to")]
+    UninitializedReferenceValueLoaded { value: String },
+    #[error(
+        "Mismatched types in binary operator: `{operator} {lhs_id}, {rhs_id}`  ({operator} {lhs}, {rhs})"
+    )]
+    MismatchedTypesInBinaryOperator {
+        lhs_id: ValueId,
+        lhs: String,
+        operator: BinaryOp,
+        rhs_id: ValueId,
+        rhs: String,
+    },
+    #[error("Division by zero: `div {lhs_id}, {rhs_id}`  ({lhs} / {rhs})")]
+    DivisionByZero { lhs_id: ValueId, lhs: String, rhs_id: ValueId, rhs: String },
+    #[error("Unsupported operator `{operator}` for type `{typ}`")]
+    UnsupportedOperatorForType { operator: &'static str, typ: &'static str },
+    #[error(
+        "Invalid bit size of `{bit_size}` given to truncate, maximum size allowed for unsigned values is {MAX_UNSIGNED_BIT_SIZE}"
+    )]
+    InvalidUnsignedTruncateBitSize { bit_size: u32 },
+    #[error(
+        "Invalid bit size of `{bit_size}` given to truncate, maximum size allowed for signed values is {MAX_SIGNED_BIT_SIZE}"
+    )]
+    InvalidSignedTruncateBitSize { bit_size: u32 },
+    #[error("Rhs of `{operator}` should be a u32 but found `{rhs_id} = {rhs}`")]
+    RhsOfBitShiftShouldBeU32 { operator: &'static str, rhs_id: ValueId, rhs: String },
+    #[error(
+        "Expected {expected_type} value in {instruction} but instead found `{value_id} = {value}`"
```

Comment:
```suggestion
        "Expected `{expected_type}` value in {instruction} but instead found `{value_id} = {value}`"
```

Reply:
"expected_type" isn't always an exact type here so I don't want to quote it. E.g. it may be `"Expected numeric value" or "Expected array or slice value"